### ISSUE TITLE
A Differentiable Reacting Flow Simulation Package in PyTorch

### ISSUE THREAD

**Issue by jiweiqi** [Cantera/enhancements#49]
**Abstract**

We are developing a differentiable reacting flow simulation package in PyTorch. We call it [ReacTorch](https://github.com/DENG-MIT/reactorch) and it is available on [Github](https://github.com/DENG-MIT/reactorch). The capability of Auto-differentiation enable us to efficiently compute the derivative of the solutions to all of the species concentrations (Jacobian) as well model parameters (Sensitivity analysis) at almost no cost. It also natively support of GPU with PyTorch. In addition, the compatibility of differentiate the whole combustion model is the foundations of many latest hybrid physics-neural network algorithms. This package will provide a easy-accessible platform for implementing those emerging hardware and software infrastructure from deep learning community.

Take advantage of the easily-parse yaml mechanism format in Cantera, we have rapidly developed a protocol with zero-D and one-D reactors. Although, the feature is not suited to be merged into Cantera, I though it is good to post here to see if there is someone else working on the same thing. We can team up to work on this "big" project.

**Comment by bryanwweber** [comment]
Thanks @jiweiqi! Sounds really cool. I'm glad you're able to take advantage of YAML, that's one of the reasons we wanted to introduce that format. If there are any changes that would make your work easier, feel free to propose them

**Comment by jiweiqi** [comment]
Hi Bryan @jiweiqi , one thing that I think it could be improved is to standardize the units across different mechanisms, at least when we use the ctml2yaml convert. Some thing annoying is to handling kcal/mol and cal/mol. Although legacy mechanism files are inconsistent, we can make it consistent in the new format of YAML :)

**Comment by bryanwweber** [comment]
Hi @jiweiqi I wish we could standardize everything, but alas, we can't control that. There are also some good benefits of matching the input to the output, such as being able to directly compare and make sure the output is correct. What we might be able to do is provide a `--convert-units` option that puts everything into SI units (+ kmol) or something like that.

**Comment by jiweiqi** [comment]
We have recently implemented the governing equations of one-dimension flame, i.e., feed-in the states at each location and compute the residuals. In addition, it can compute the derivatives of the residuals with respect to the state variables (T, Y, u) and all of the parameters (pre-factor A, b, Ea, thermo data, and transport data). **The computational cost should be approximately independent of the number of grid points, the number of state variables, and the number of parameters since they are evaluated in batch on GPU.** This feature might be interested to explore with coupling with Cantera, as an extension to current `solve_adjoint`. For instance, with those derivatives, it is straightforward to compute the sensitivity of mass fraction/temperature at each grid point and the parameters include more than pre-factor A. The computation starts from the converged solution from Cantera since we are not going to implement solvers for the one-dimensional flame.

BTW, is there any need to improve the performance of Cantera in terms of flame speed sensitivity analysis? I guess one of the bottlenecks could be the computation of the severities of the residuals to parameters.

Hi, @speth @kyleniemeyer any suggestion on this proposal?

**Comment by jiweiqi** [comment]
Hi @speth, regarding the function of `solve_adjoint`, if I understand it correctly, it is implementing the following equation, right?
![image](https://user-images.githubusercontent.com/8445647/93125167-bcba2880-f698-11ea-9ee0-8621becde146.png)
The mage if taken from Chemkin manual https://personal.ems.psu.edu/~radovic/ChemKin_Theory_PaSR.pdf.

**Comment by ischoegl** [comment]
@jiweiqi ... I believe some of the rationale you provide in DENG-MIT/reactorch/issues/5 falls short. Reimplementing the 'second category' is de facto reimplementing the core of what Cantera intends to be (cantera means 'quarry' in spanish, and there really is the intent to provide 'buidling blocks' to the research community). The first category you mention are really examples for what can be done with Cantera. Without the second, there is no first.

**Comment by jiweiqi** [comment]
Hi @ischoegl , we may have to limit the scope of the discussions within deep learning for combustion research. You may hear of PINN and the development of ReacTorch is largely motivated for algorithms like PINN. The package put everything in the second category into PyTorch, which facilitates the training of coupled neural networks. We are still working on it since there is some fundamental challenges of training PINN for stiff systems as outlined in this [paper](https://arxiv.org/abs/2011.04520).

I agree if discussed under the context of conventional CFD, there is no need to reinvent the second category as well. For instance, one can fully rely on Cantera in OpenFoam.

So the major motivation is to augment chemical kinetic models into deep learning frameworks. Developing new packages is very painful especially for our newly established lab. The first choice will always be finding existing tools. If there are other ways to do it, definitely will try.

By the way, the implementation in ReacTorch is not from scratch, it retrieves information from Cantera whenever there is an exposed API.

**Comment by ischoegl** [comment]
@jiweiqi ... I don’t think that needs of machine learning and more ‘traditional’ approaches are as different as you make them sound. If you are willing to dig a little deeper, all sides ultimately need the exact same tools that you decided aren’t worth pursuing within what you perceive as a ‘legacy’ framework. Your preference is certainly faster for short term goals, and I understand that the lab is new. Keeping it going for the long haul will be the real task.

**Comment by jiweiqi** [comment]
Hi @ischoegl , we could limit the discussions on the example of PINN. Could you echo more on how to do it in Cantera or Chemkin?

**Comment by ischoegl** [comment]
To mention one of the aspects of your manuscript, Quasi-Steady-State-Assumptions (QSSA) aren’t specific to PINN (nor is the concept new, see work at UConn). To your point: if it’s not part of Cantera (which it isn’t at the moment), be part of the discussion and just pitch in. In return, you’ll be able to leverage others’ work in the long run.

**Comment by jiweiqi** [comment]
@ischoegl , we can have the discussions on the paper in other places. Likewise, QSSA is something existing over a century. The motivation of the work is to make PINN works for stiff systems, and QSSA is one potential solution. It is more of an example of using classical combustion knowledge to solve machine learning problems.

Regarding the point enhancement of Cantera, we should definitely discuss the possibility of extending Cantera for it. I can share more when more people in the combustion community start looking into PINN and propose enhancements to Cantera. AD is certainly the requirement. What I am not sure of is how many additional features required to enable PINN in Cantera.

**Comment by rwest** [comment]
I don't want to squash helpful discussion, but would like to make a reminder that we expect everyone to contribute to creating a positive environment, so even bystanders and others reading but not participating in the discussion feel like it's a welcoming place. Examples include:
- Being respectful of differing viewpoints and experiences
- Gracefully accepting constructive criticism
- Focusing on what is best for the community
- Showing empathy towards other community members
Thanks!

**Comment by jiweiqi** [comment]
@ischoegl , I am very grateful for your insights. Forgive me if any part of my post is in an improper tongue. As a non-native speaker, it is pretty often for me to use an inaccurate and inproper tongue without realizing it.

**Comment by ischoegl** [comment]
@jiweiqi ... no problem, I do see that we have differing viewpoints as @rwest clearly recognized. I never intended to sound harsh myself (but am not a native speaker either). I hope you know that I believe your work has merit despite our disagreements.

**Comment by ischoegl** [comment]
@jiweiqi ... truly happy that you opened #82 - I was hoping to see a discussion along those lines. Regarding your comments about gaps in Cantera's Python API that you may have referenced to in #81, please let me know specifics. I believe there's a lot of value to make sure that all kinetic parameters are properly exposed, so a re-interpretation of a YAML input file should not be necessary within ReacTorch (the value is not specific to ML, and the entire community will benefit). I know the code-base well, and exposing parameters available in C++ to the Python front end is not too difficult.

**Comment by jiweiqi** [comment]
I will bother you when we come back to upgrading ReacTorch's interpreter.

For now, I have a quick question about the API. Is there an easy way to get the elementary matrix? basically, how many atoms of each elements in a molecule. For example, for species [CH4, H2, H2O], it will be
[1, 0, 0] C
[3, 2, 2] H
[0, 0, 1] O
It is not difficult to extract it from YAML, but it will be great if I can access it in Python.

**Comment by ischoegl** [comment]
Not in a single call, but it's not difficult:
```python
[([gas.n_atoms(sp, el) for sp in ['CH4', 'H2', 'H2O']], el)  for el in 'CHO']
```

**Comment by jiweiqi** [comment]
That's a very clever way!

**Comment by ischoegl** [comment]
@jiweiqi ... I've used it before ... even for the same purpose, but not within the context of ML.

**Comment by ischoegl** [comment]
For accessing `Reaction` kinetics parameters from Python, here are some examples:
```python
gas = ct.Solution('gri30.yaml')

rxn = gas.reaction(2) # Reaction 2 is an elementary reaction
rr = rxn.rate # Arrhenius class holding rate parameters
A, b, e = rr.pre_exponential_factor, rr.temperature_exponent, rr.activation_energy

rxn = gas.reaction(0) # Reaction 0 is a threebody reaction  
rxn.efficiencies # third-body-efficiencies
rr = rxn.rate # Arrhenius class holding rate parameters

rxn = gas.reaction(303) # Reaction 303 is a falloff reaction
rr = rxn.low_rate # Arrhenius class holding parameters
rr = rxn.high_rate # Arrhenius class holding parameters
```
I'm not certain that everything is accessible, but as I said it's relatively straightforward to add to the API.

**Comment by jiweiqi** [comment]
@ischoegl  does the syntax of `gas.reaction(2).rate` works? I didn't know the class of rxn.rate, which seems not being documented. I should have checked the source code ...

**Comment by ischoegl** [comment]
That will work, just like `gas.reaction(2).rate.pre_exponential_factor`. I haven’t checked the docs, but yes, it’s not as obvious as other parts.

**Comment by speth** [comment]
The `rate` member of the `ElementaryReaction` class is documented as you can see here: https://cantera.org/documentation/docs-2.5/sphinx/html/cython/kinetics.html#elementaryreaction.

I might suggest that questions about how to use Cantera are a better fit for the Users' Group than threads attached to this or any other enhancement post.

**Comment by bryanwweber** [comment]
@jiweiqi For the same reasons as #89, I'm going to close this issue as well. If there are specific changes that are needed in Cantera, please update the issue. Thanks!
